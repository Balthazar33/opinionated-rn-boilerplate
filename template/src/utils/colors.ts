export enum Colors {
  WHITE = '#ffffff',
  BLACK = '#000000',
  PRIMARY_RED = '#ff0000',
  PRIMARY_BLUE = '#0000ff',
  PRIMARY_GREEN = '#00ff00',
  GREY = '#8a8a8a',
  TRANSPARENT = 'transparent',
}

// Opacity range type implementation----------------------
// Ref: https://stackoverflow.com/a/70307091/14431966
type Enumerate<
  N extends number,
  Acc extends number[] = [],
> = Acc['length'] extends N
  ? Acc[number]
  : Enumerate<N, [...Acc, Acc['length']]>;
type Range<F extends number, T extends number> = Exclude<
  Enumerate<T>,
  Enumerate<F>
>;
type OpacityRange = Range<0, 100>;
//--------------------------------------------------------

/**
 * Returns a color with the specified opacity
 * @return color with the specified opacity
 */
export const withOpacity = (color: string, opacity: OpacityRange) => {
  return `${color}${opacity}`;
};
