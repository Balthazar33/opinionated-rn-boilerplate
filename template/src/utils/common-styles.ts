import {StyleSheet} from 'react-native';

export const commonStyles = StyleSheet.create({
  flex1: {
    flex: 1,
  },
  flexGrow1: {
    flexGrow: 1,
  },
  justifyAndAlignCenter: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  rowCenter: {
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  margin0: {
    marginTop: 0,
    marginLeft: 0,
    marginRight: 0,
    marginBottom: 0,
  },
});
