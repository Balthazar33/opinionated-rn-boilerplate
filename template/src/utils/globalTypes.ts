// Suffixes for RTK query status
export enum RequestState {
  PENDING = 'pending',
  FULFILLED = 'fulfilled',
  REJECTED = 'rejected',
}
