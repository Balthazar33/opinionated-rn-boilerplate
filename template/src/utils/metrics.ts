import {PixelRatio, Dimensions} from 'react-native';

const {width: screenWidth, height: screenHeight} = Dimensions.get('screen');

/**
 * Returns device-adjusted size for the given input
 * @param {number} input - input size
 * @returns responsive size
 * References:
 * 1. https://medium.com/@HelloMoto69/react-native-responsive-scaling-font-dimensions-pixels-5dcccd8f7124#:~:text=By%20dividing%20the%20device's%20screen,screen%20size%20and%20aspect%20ratio.
 * 2. https://reactnative.dev/docs/pixelratio#roundtonearestpixel
 */
export const sizer = (input: number) => {
  return PixelRatio.roundToNearestPixel((input * screenWidth) / 375);
};

export {screenHeight, screenWidth};
