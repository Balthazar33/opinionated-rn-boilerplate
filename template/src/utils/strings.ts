export const Strings = {
  noInternet: 'No Internet Connection',
  error400: 'Incorrect Data',
  error401: 'Unauthorized action',
  error404: 'Resource not found',
  error500: 'Internal server error',
  genericError: 'Something went wrong',
  fetchError: 'fetch_error',
  networkRequestFailed: 'network request failed',

  // dummy screen strings
  baseScreen: 'Base Screen',
  baseScrollScreen: 'Base Scroll Screen',
  toApiCallScreen: 'To Api Call Screen',
  apiCallScreen: 'Api Call Screen',
  getAllPokemon: 'Get All Pokemon',

  throwError: 'Throw Error',
  whoops: 'Whoops!',
  tryAgain: 'Try Again',
  ok: 'OK',
  error: 'Error',
  username: 'Username',
  contactNo: 'Contact No.',
  email: 'Email',
  contact: 'Contact',
};
