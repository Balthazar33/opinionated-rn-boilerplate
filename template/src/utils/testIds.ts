export const TestIds = {
  APP_SAFEAREA: 'AppSafeArea',
  APP_ACTIVITY_INDICATOR: 'ActivityIndicator',
  NO_INTERNET: 'NoInternet',
};
