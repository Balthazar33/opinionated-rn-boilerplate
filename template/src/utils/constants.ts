import {sizer} from './metrics';

export const SCREEN_PADDING = sizer(16);
