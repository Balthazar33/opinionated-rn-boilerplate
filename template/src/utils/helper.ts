import {Platform} from 'react-native';
import {Strings} from './strings';

export const isAndroid = Platform.OS === 'android';

// Determine if the error is network-related
export const isNetworkError = (error: any) => {
  return (
    error?.status?.toString().toLowerCase() === Strings.fetchError ||
    error?.errror
      ?.toString()
      .toLowerCase()
      .includes(Strings.networkRequestFailed)
  );
};
