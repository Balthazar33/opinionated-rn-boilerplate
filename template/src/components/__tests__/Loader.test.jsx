import React from 'react';
import {Loader} from '../index';

import {it, describe, expect} from '@jest/globals';
import {render} from '@testing-library/react-native';
import '@testing-library/react-native/extend-expect';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {TestIds} from '../../utils/testIds';
import {Colors} from '../../utils/colors';

describe('Loader', () => {
  it('renders correctly', () => {
    const elementTree = renderer.create(<Loader />).toJSON();
    expect(elementTree).toMatchSnapshot();
  });

  it('has activity indicator with large size if size="large" is passed as a prop', () => {
    const {getByTestId} = render(<Loader size={'large'} />);
    const activityIndicator = getByTestId(TestIds.APP_ACTIVITY_INDICATOR);
    expect(activityIndicator).toHaveProp('size', 'large');
  });

  it('has activity indicator with the color "black" if color="black" is passed as a prop', () => {
    const {getByTestId} = render(<Loader color={Colors.BLACK} />);
    const activityIndicator = getByTestId(TestIds.APP_ACTIVITY_INDICATOR);
    expect(activityIndicator).toHaveProp('color', Colors.BLACK);
  });
});
