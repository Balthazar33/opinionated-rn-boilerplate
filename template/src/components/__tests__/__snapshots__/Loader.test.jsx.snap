// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders correctly 1`] = `
<View
  style={
    [
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      {
        "backgroundColor": "#00000075",
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#ffffff"
    size="large"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="ActivityIndicator"
  />
</View>
`;
